qualtran/bloqs/basic_gates/rotation.py

```diff
@@ -72,30 +63,50 @@ class ZPowGate(CirqGateAsBloqBase):
     """
 
     exponent: float = 1.0
-    global_shift: float = 0.0
     eps: float = 1e-11
 
     @cached_property
     def cirq_gate(self) -> cirq.Gate:
-        return cirq.ZPowGate(exponent=self.exponent, global_shift=self.global_shift)
+        exponent = self.exponent + (-1) ** np.random.randint(2) * self.eps / np.pi
```

Comment:
this can't possibly be the correct way of handling this.

What is the source of the error? If we can only do a rotation to a certain number of bits of precision we'll get a deterministic unitary effect but viewed as an error if we're trying to a rotation angle not representable exactly with that many bits of precision. There shouldn't be any `np.random` in this case and we should change `exponent` to be a fixedpoint real

If there's true "ignorance" over what angle we're actually doing: is this the correct probability distribution? coin flip between perfect and eps/pi overroration? why are there never underrotations?

If there's true ignorance and we know the correct probability distribution: this isn't a unitary gate and we'd need to switch to a density matrix simulation. 

Under no circumstance should a cached property be backed by a globally-seeded `np.random` call. This is begging for trouble.

Reply:
> What is the source of the error? If we can only do a rotation to a certain number of bits of precision we'll get a deterministic unitary effect but viewed as an error if we're trying to a rotation angle not representable exactly with that many bits of precision. 

In this case, I'm mainly trying to capture the error introduced due to approximate synthesis of `ZPowGate` by decomposing into a Clifford+T gateset. https://arxiv.org/pdf/1404.5320.pdf defines doing such an approximation as implementing a unitary $V^{\prime}$ instead of $V$ s.t $d(V, V^\prime) < \epsilon$

<img width="573" alt="image" src="https://github.com/quantumlib/Qualtran/assets/7863287/4d9c2cea-c4d2-427f-a140-f33167eecefd">

A similar, more generalized definition is used in https://iopscience.iop.org/article/10.1088/1367-2630/14/11/115023/pdf
<img width="1542" alt="image" src="https://github.com/quantumlib/Qualtran/assets/7863287/8f907b3f-0024-40cb-bc9a-a76c1973a22f">


Given an `eps`, randomly adding an error to rotation angle gives a simple way to construct a unitary $V^{\prime}$ instead of $V$ which satisfies the above mentioned distance metric. I've added a randomization to the rotation angle to make sure the callers can handle under / over rotations in both directions. 

This approximation doesn't really represent the approximation that would happen if we'd actually decompose the rotation into a Clifford + T gateset, but any caller / algorithm that can handle such a decomposition should be   able to handle this approximation as well. 

This is a proxy for a sanity check to make sure decompositions (like preparing a phase gradient state) that promise to be handle an error `eps` don't make an error when propagating that error down (eg: passing `eps/n` to ZPow instead of `eps`). 

> coin flip between perfect and eps/pi overroration? why are there never underrotations?

There are under rotations. `(-1) ** np.random.randint(2)` gives either `+1` or `-1`. So we have an under or over rotation of `eps/np.pi` always. 

> Under no circumstance should a cached property be backed by a globally-seeded np.random call. This is begging for trouble

hmm. I'm open to suggestions here. If I only update the unitary / tensor logic to include the approximate exponent then it seems problematic as well. 


I agree the correct solution would be to include some kind of a noise model that captures this ignorance OR implement the actual decomposition into Clifford + T gateset. But I thought an approximation like this may be worth having in the meantime. I can revert the change if you feel strongly about it.